Instruments/python/btag_norm_sf.py

```diff
@@ -0,0 +1,177 @@
+    "TauVSjetSF_pt35to40": 106, "TauVSjetSF_ptgt40": 107, "EleFakingTauES_DM0": 108, "EleFakingTauES_DM1": 109,
+    "EleFakingTauES_3prong": 110, "TauVSeSF_barrel": 111, "TauVSeSF_endcap": 112, "MuFakingTauES_DM0": 113,
+    "MuFakingTauES_DM1": 114, "MuFakingTauES_3prong": 115, "TauVSmuSF_etaLt0p4": 116, "TauVSmuSF_eta0p4to0p8": 117,
+    "TauVSmuSF_eta0p8to1p2": 118, "TauVSmuSF_eta1p2to1p7": 119, "TauVSmuSF_etaGt1p7": 120, "MuFakingTauES": 121,
+    "EleIdIsoUnc": 122, "MuonIdIsoUnc": 123, "TauTriggerUnc_DM0": 124, "TauTriggerUnc_DM1": 125, "TauTriggerUnc_DM10": 126,
+    "TauTriggerUnc_DM11": 127, "L1_prefiring": 128, "PileUp": 129, "PileUpJetId_eff" : 130, "PileUpJetId_mistag" : 131,
+    "TauCustomSF_DM0": 132, "TauCustomSF_DM1": 133, "TauCustomSF_DM10": 134, "TauCustomSF_DM11": 135,
+    "VBFTriggerUnc_jets": 139, "VBFTauTriggerUnc_DM0": 140, "VBFTauTriggerUnc_DM1": 141, "VBFTauTriggerUnc_3prong": 142,
+    "btag_lf": 143, "btag_hf": 144, "btag_hfstats1": 145, "btag_hfstats2": 146,
+    "btag_lfstats1": 147, "btag_lfstats2": 148, "btag_cferr1": 149, "btag_cferr2": 150}
+variation_dictionary = { "Central" : 0, "Up" : 1, "Down" : -1 }
+
+
+btag_unc_sources=["btag_lf","btag_hf","btag_hfstats1","btag_hfstats2","btag_lfstats1","btag_lfstats2"]#,"btag_cferr1","btag_cferr2"]
+jes_unc_sources=["JetReduced_Absolute","JetReduced_Absolute_year","JetReduced_BBEC1","JetReduced_BBEC1_year","JetReduced_EC2","JetReduced_EC2_year","JetReduced_FlavorQCD","JetReduced_HF","JetReduced_HF_year","JetReduced_RelativeBal","JetReduced_RelativeSample_year","JetReduced_Total"]
+les_unc_sources=["TauES", "TauES_DM0", "TauES_DM1", "TauES_DM10", "TauES_DM11", "EleFakingTauES_DM0", "EleFakingTauES_DM1", "MuFakingTauES"]
+unc_variations=['Up', 'Down']
+
+def add_dic(r_factor, unc_source,unc_scale, r_factors_dict):
+    print ("r factor related to %s and %s  is %.10f " % (unc_source, unc_scale,r_factor))
+    if isnan(r_factor):
+        r_factor = 1
+    r_factors_dict[unc_source][unc_scale]=r_factor
+
+def evaluate_r_Central(file,channel, unc_sources, unc_scales, r_factors_dict, isTune5):
+    #print file
+    d = ROOT.RDataFrame(channel, file)
+    not_data = '! is_data'
+    d=d.Filter(not_data)
+    if isTune5 == 1:
+        d=d.Filter("is_TuneCP5==1")
+    elif isTune5 == 2:
+        d=d.Filter("is_TuneCP5==0")
+    for unc_source in unc_sources:
+        for unc_scale in unc_scales:
+            num = "num_"+unc_source+"_"+unc_scale
+            den = "den_"+unc_source+"_"+unc_scale
+            weight_num = "weight"
+            if unc_source=="None":
+                weight_den = "weight*weight_btag_IterativeFit"
+            else:
+                weight_den = "weight*weight_btag_IterativeFit*unc_"+unc_source+"_"+unc_scale
+            w_central_before = d.Define(num, weight_num).Sum(num)
+            w_central_after = d.Define(den, weight_den).Sum(den)
+            numerator= w_central_before.GetValue()
+            denominator= w_central_after.GetValue()
```

Comment:
- don't call GetValue here: store w_central_before and w_central_after and compute r_factor later, after all unc are collected.
- there is many duplicated logic between evaluate_r_Central and evaluate_r_JESLES -> please refactor the code.